test/Microsoft.Azure.Devices.Edge.Test/helpers/Context.cs

```diff
@@ -0,0 +1,161 @@
+                    Preconditions.CheckArgument(File.Exists(key));
+                    return Option.Some((certificate, key, password));
+                }
+
+                return Option.None<(string, string, string)>();
+            }
+
+            string GenerateDefaultDeviceId() =>
+                $"e2e-{string.Concat(Dns.GetHostName().Take(15))}-{DateTime.Now:yyMMdd'-'HHmmss'.'fff}";
+
+            const int DeviceIdMaxLength = 37;
+            string CheckDeviceIdLength(string deviceId)
+            {
+                if (deviceId.Length > DeviceIdMaxLength)
+                {
+                    string message =
+                        $"The device ID is more than {DeviceIdMaxLength} characters in length, and " +
+                        "may cause problems in tests that generate certificates with a CN based on " +
+                        "the device ID.";
+                    throw new ArgumentException(message);
+                }
+
+                return deviceId;
+            }
+
+            this.CaCertScriptPath = Get("caCertScriptPath");
+            this.ConnectionString = Get("IOT_HUB_CONNECTION_STRING");
+            this.DeviceId = CheckDeviceIdLength(GetOrDefault("deviceId", GenerateDefaultDeviceId()));
+            this.EdgeAgentImage = Option.Maybe(Get("edgeAgentImage"));
+            this.EdgeHubImage = Option.Maybe(Get("edgeHubImage"));
+            this.EventHubEndpoint = Get("EVENT_HUB_ENDPOINT");
+            this.InstallerPath = Option.Maybe(Get("installerPath"));
+            this.LogFile = Option.Maybe(Get("logFile"));
+            this.MethodReceiverImage = Option.Maybe(Get("methodReceiverImage"));
+            this.MethodSenderImage = Option.Maybe(Get("methodSenderImage"));
+            this.OptimizeForPerformance = context.GetValue("optimizeForPerformance", true);
+            this.PackagePath = Option.Maybe(Get("packagePath"));
+            this.Proxy = Option.Maybe(context.GetValue<Uri>("proxy"));
+            this.Registries = GetAndValidateRegistries();
+            this.RootCaKeys = GetAndValidateRootCaKeys();
+            this.SetupTimeout = TimeSpan.FromMinutes(context.GetValue("setupTimeoutMinutes", 5));
+            this.TeardownTimeout = TimeSpan.FromMinutes(context.GetValue("teardownTimeoutMinutes", 2));
+            this.TempSensorImage = Option.Maybe(Get("tempSensorImage"));
+            this.TestTimeout = TimeSpan.FromMinutes(context.GetValue("testTimeoutMinutes", 5));
+            this.Verbose = context.GetValue<bool>("verbose");
+        }
+
+        static readonly Lazy<Context> Default = new Lazy<Context>(() => new Context());
+
+        public static Context Current => Default.Value;
+
+        public string CaCertScriptPath { get; }
+
+        public string ConnectionString { get; }
+
+        public string DeviceId { get; }
+
+        public Option<string> EdgeAgentImage { get; }
+
+        public Option<string> EdgeHubImage { get; }
```

Comment:
why EdgeHub image is optional? Although edgent agent image is defined in config.yaml; is it clear to make edgeAgent image required?

Reply:
Edge Hub should be optional because it isn't strictly necessary; you can run/test modules without it. Also we want to be able to fall back to the mcr.microsoft.com image if the tests need Edge Hub but context.json doesn't specify it.

Edge Agent is required, but like Edge Hub we want to use the default mcr.microsoft.com image if you don't specify it.

However, this raises an issue: even though the `ModuleConfigBuilder` code uses the default image for Edge Agent and Hub if you don't give one, the existing tests all force you to provide one in context.json anyway.

I just added a commit that lets the tests use the default images if you don't specify them in context.json. Now, you can run the TempSensor test with an empty context.json and only two environment variables: the IoT Hub connection string and Event Hub-compatible endpoint.